### PR TITLE
[SPARK-48555][SQL][PYTHON][CONNECT] Support using Columns as parameters for several functions in pyspark/scala

### PR DESCRIPTION
**Pull request by Ronserruya**
### What changes were proposed in this pull request?

https://issues.apache.org/jira/browse/SPARK-48555  
For pyspark, added the ability to use the "Column" type or names of column for the parameters of the following functions:

- array_remove
- array_position
- map_contains_key
- substring

For scala, added the ability to use "Column" type as the parameters of the `substring` function

This functionality already exists in the SQL syntax:
```
select array_remove(col1, col2) from values (array(1,2,3), 2)
```

however, it isn't possible to do the same in python
```python
df.select(F.array_remove(F.col("col1"), F.col("col2"))
```

Note that in scala the functions other than `substring` already accepted Column params (or rather, they accept `Any` and pass whatever the param is to `lit` so it ends up working), so I  only needed to change substring in the scala side. 

### Why are the changes needed?
To align the scala/python API with the SQL one.

### Does this PR introduce _any_ user-facing change?
Yes, added new overloaded functions in scala and changed type hints/docstrings in python.

### How was this patch tested?
Added doctests for the python changes, and tests in the scala test suites, then tested both manually and using the CI.

### Was this patch authored or co-authored using generative AI tooling?
No.

### Notes:

- I opened the related JIRA ticket, but looks like I don't have the option to assign it myself, so if it is required and any reviewer does have permissions for it, I'd appreciate it
- The "Build" workflow passed successfully, but the "report tes results" one didn't due to some authorization issue, I see that this is the same for many other open PRs right now so I assume its ok.
- For the python changes, I tried to follow the convention used by other functions (such as `array_contains` or `when`), of using `value._jc if isinstance(value, Column) else value`
- Im not really familiar with the `connect` functions, but seems like on the python side they already supported the use of columns so no extra changes were needed there
- For the scala side, this is the first time I'm touching scala, I think I covered it all as I tried to match similar changes done in [a similar PR](https://github.com/apache/spark/pull/46045)
- The same issue also exists for `substring_index` however I wasn't able to fix this one the same way I did for `substring`. Calling it with a `lit` for the `count` arg worked, but using a `col` error with a `NumberFormatError` for "project_value_3". I assume the error is related to trying to parse the Int [here](https://github.com/apache/spark/blob/7cba1ab4d6acef4e9d73a8e6018b0902aac3a18d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/stringExpressions.scala#L1449) . In that case I got lost in the scala code and decided to drop it, but if anyone knows how to fix this error I could change that function as well.

The contribution is my original work and I license the work to the project under the project’s open source license.

### CODE REVIEW

**Comment by Ronserruya** [comment]
tagging @CTCC1 @LuciferYang @zhengruifeng As you wrote/reviewed the similar PR I mentioned 😄

File: connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala [hunk 1]
```diff
@@ -4276,6 +4276,45 @@ object functions {
...
+   * of byte array that starts at `pos` in byte and is of length `len` when str is Binary type
+   *
+   * @note
+   *   The position is not zero based, but 1 based index.
+   *
+   * @group string_funcs
+   * @since 4.0.0
+   */
+  def substring(str: Column, pos: Int, len: Column): Column =
+    Column.fn("substring", str, lit(pos), len)
```
Comment (zhengruifeng) [inline connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala:4303]:
```suggestion

```

let's only add signature `def substring(str: Column, pos: Column, len: Column): Column`

referring to https://github.com/apache/spark/blob/a7da9b6b8aed99c5df23ccd83ab21a4a9a50d28a/connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/functions.scala#L55-L57

File: python/pyspark/sql/functions/builtin.py [hunk 2]
```diff
@@ -10949,9 +10954,20 @@ def substring(str: "ColumnOrName", pos: int, len: int) -> Column:
...
+    >>> df.select(substring(df.s, df.p, 3).alias('s')).collect()
+    [Row(s='par')]
+    >>> df.select(substring(df.s, df.p, df.l).alias('s')).collect()
+    [Row(s='par')]
+    """
+    # the native str type is shadowed by the function's `str` param
+    from builtins import str as StrType
     from pyspark.sql.classic.column import _to_java_column
 
+    pos = _to_java_column(pos) if isinstance(pos, (StrType, Column)) else pos
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions/builtin.py:10969]:
```suggestion
    pos = pos if isinstance(pos, int) else _to_java_column(pos)
```
then we don't need `StrType`

File: python/pyspark/sql/functions/builtin.py [hunk 3]
```diff
@@ -14034,9 +14053,22 @@ def array_position(col: "ColumnOrName", value: Any) -> Column:
...
+    +-------------------------+
+    |array_position(data, col)|
+    +-------------------------+
+    |                        2|
+    +-------------------------+
+
     """
     from pyspark.sql.classic.column import _to_java_column
 
+    value = value._jc if isinstance(value, Column) else value
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions/builtin.py:14071]: doesn't `_to_java_column` already support `Column` type input?

File: python/pyspark/sql/functions/builtin.py [hunk 4]
```diff
@@ -14470,9 +14505,21 @@ def array_remove(col: "ColumnOrName", element: Any) -> Column:
...
+    >>> df.select(sf.array_remove(df.data, df.col)).show()
+    +-----------------------+
+    |array_remove(data, col)|
+    +-----------------------+
+    |                 [2, 3]|
+    +-----------------------+
     """
     from pyspark.sql.classic.column import _to_java_column
 
+    element = element._jc if isinstance(element, Column) else element
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions/builtin.py:14522]: ditto

File: python/pyspark/sql/functions/builtin.py [hunk 5]
```diff
@@ -17267,9 +17317,21 @@ def map_contains_key(col: "ColumnOrName", value: Any) -> Column:
...
+    >>> df.select(sf.map_contains_key("data", sf.col("key"))).show()
+    +---------------------------+
+    |map_contains_key(data, key)|
+    +---------------------------+
+    |                       true|
+    +---------------------------+
     """
     from pyspark.sql.classic.column import _to_java_column
 
+    value = value._jc if isinstance(value, Column) else value
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions/builtin.py:17334]: ditto

Comment (zhengruifeng) [reply python/pyspark/sql/functions/builtin.py:17334]: The `Example 3: Check for key using a column` was already supported

File: sql/core/src/main/scala/org/apache/spark/sql/functions.scala [hunk 6]
```diff
@@ -4234,6 +4234,45 @@ object functions {
...
+   * when str is Binary type
+   *
+   * @note The position is not zero based, but 1 based index.
+   *
+   * @group string_funcs
+   * @since 4.0.0
+   */
+  def substring(str: Column, pos: Int, len: Column): Column =
+    Column.fn("substring", str, lit(pos), len)
+
```
Comment (zhengruifeng) [inline sql/core/src/main/scala/org/apache/spark/sql/functions.scala:4262]:
```suggestion

```

File: python/pyspark/sql/functions/builtin.py [hunk 7]
```diff
@@ -10926,6 +10928,9 @@ def substring(str: "ColumnOrName", pos: int, len: int) -> Column:
     .. versionchanged:: 3.4.0
         Supports Spark Connect.
 
+    .. versionchanged:: 4.0.0
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions/builtin.py:10931]: let's move it under parameters section, you may refer to

https://github.com/apache/spark/blob/bb1f026da0e3193e0cfad2001c7a8cab59a5290a/python/pyspark/sql/functions/builtin.py#L6216-L6220

File: python/pyspark/sql/functions/builtin.py [hunk 8]
```diff
@@ -13959,6 +13973,9 @@ def array_position(col: "ColumnOrName", value: Any) -> Column:
...
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions/builtin.py:13976]: ditto

File: python/pyspark/sql/functions/builtin.py [hunk 9]
```diff
@@ -14397,12 +14427,15 @@ def array_remove(col: "ColumnOrName", element: Any) -> Column:
...
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions/builtin.py:14430]: ditto

File: python/pyspark/sql/functions/builtin.py [hunk 10]
```diff
@@ -17232,12 +17277,15 @@ def map_contains_key(col: "ColumnOrName", value: Any) -> Column:
...
```
Comment (zhengruifeng) [inline python/pyspark/sql/functions/builtin.py:17280]: ditto

**Review by zhengruifeng — APPROVED** [review]

File: python/pyspark/sql/functions/builtin.py (hunk 5 above)
Comment (zhengruifeng) [reply python/pyspark/sql/functions/builtin.py]: I see, it was not supported in Classic mode, but supported in Connect mode.

Classic:
```
In [2]: df = spark.sql("select map(1, 2, 3, 4) as m, 1 as k")
   ...: df.select(F.map_contains_key(df.m, df.k))
---------------------------------------------------------------------------
PySparkTypeError                          Traceback (most recent call last)
Cell In[2], line 2
      1 df = spark.sql("select map(1, 2, 3, 4) as m, 1 as k")
----> 2 df.select(F.map_contains_key(df.m, df.k))

...

File ~/Dev/spark/python/pyspark/sql/classic/column.py:415, in Column.__iter__(self)
    414 def __iter__(self) -> None:
--> 415     raise PySparkTypeError(
    416         error_class="NOT_ITERABLE", message_parameters={"objectName": "Column"}
    417     )

PySparkTypeError: [NOT_ITERABLE] Column is not iterable.

```

Connect:
```
In [1]: from pyspark.sql import functions as F

In [2]: df = spark.sql("select map(1, 2, 3, 4) as m, 1 as k")
   ...: df.select(F.map_contains_key(df.m, df.k))
Out[2]: DataFrame[map_contains_key(m, k): boolean]
```

There is a slight difference in the handling of `value: Any` typed value: Spark Connect always convert `value: Any` to Column, while some functions (e.g. `map_contains_key`) in Classic don't do this.

We will need to revisit all the `Any` typed parameters in functions. cc @HyukjinKwon

**Comment by zhengruifeng** [comment]
LGTM, only need a few minor doc changes. Thanks for working on this.

File: python/pyspark/sql/functions/builtin.py (hunk 5 above)
Comment (allisonwang-db) [reply python/pyspark/sql/functions/builtin.py]: This is an interesting behavior difference. What's the reason for not converting the classic PySpark value to a column/expression?

**Comment by Ronserruya** [comment]
@zhengruifeng Fixed the docstrings, thanks for reviewing :)
I don't see the merge bottom, so I assume I don't have the permission for it

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

File: python/pyspark/sql/functions/builtin.py (hunk 5 above)
Comment (zhengruifeng) [reply python/pyspark/sql/functions/builtin.py]: I guess it was not by design, seems just due to the type mismatch in the internal helper functions

**Pull request closed without merging**